pkg/schedulingtypes/jobscheduler.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulingtypes
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/typeconfig"
+	fedschedulingv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/scheduling/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	. "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util/planner"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+const (
+	JSPKind          = "JobSchedulingPreference"
+	ParallelismField = "parallelism"
+	CompletionsField = "completions"
+	ParallelismPath  = "spec.parallelism"
+	CompletionsPath  = "spec.completions"
+)
+
+func init() {
+	schedulingType := SchedulingType{
+		Kind:             JSPKind,
+		SchedulerFactory: NewJobScheduler,
+	}
+	RegisterSchedulingType("jobs.batch", schedulingType)
+}
+
+type JobScheduler struct {
+	plugin *Plugin
+
+	fedClient fedclientset.Interface
+
+	controllerConfig *ControllerConfig
+
+	eventHandlers SchedulerEventHandlers
+}
+
+func NewJobScheduler(controllerConfig *ControllerConfig, eventHandlers SchedulerEventHandlers) (Scheduler, error) {
+	fedClient, _, _ := controllerConfig.AllClients("job-scheduler")
+	scheduler := &JobScheduler{
+		controllerConfig: controllerConfig,
+		eventHandlers:    eventHandlers,
+		fedClient:        fedClient,
+	}
+
+	return scheduler, nil
+}
+
+func (j *JobScheduler) Kind() string {
+	return RSPKind
+}
+
+func (j *JobScheduler) StartPlugin(typeConfig typeconfig.Interface, stopChan <-chan struct{}) error {
+	kind := typeConfig.GetTemplate().Kind
+	plugin, err := NewPlugin(j.controllerConfig, j.eventHandlers, typeConfig)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize replica scheduling plugin for %q: %v", kind, err)
+	}
+	plugin.Start(stopChan)
+	j.plugin = plugin
+
+	return nil
+}
+
+func (j *JobScheduler) ObjectType() pkgruntime.Object {
+	return &fedschedulingv1a1.JobSchedulingPreference{}
```

Comment:
(No action required) This repo now vendors `controller-runtime`, so generic client/informer/lister is a thing.  Consider migrating away from the legacy generated stuff in future iterations.